torch/csrc/autograd/python_function.h

```diff
@@ -88,7 +88,7 @@ struct THPFunction {
 
     // The C++ wrapper for this Python function.
     // See a comment in THPFunction_asFunction for details about this field.
-    torch::autograd::PyFunction cdata;
+    torch::autograd::PyFunction* cdata;
```

Comment:
Why did that change to a pointer? What is the ownership semantics now? Is it strong? Is it weak?

Reply:
Still strong. A THPFunction owns the underlying PyFunction (should I make this into a std::unique_ptr to make this clearer, or add more comments?)

Here's the reason why I made it into a pointer. Let's say we have the following computation graph:

(1) shared_ptr<PyFunction> -> (2) Edge -> (3) shared_ptr<Function> -> (4) Edge -> (5) shared_ptr<Function> -> ...

When this computation graph gets freed, what first happens is that the use_count of (1) goes to 0. When that happens:
- The PyFunction doesn't get destroyed yet. Each PyFunction shared_ptr has a special destructor that does one thing: decref the python refcount on the THPFunction that owns the PyFunction.
- The THPFunction's python refcount goes to 0 if there are no more references. When that happens, the THPFunction's deleter gets called.
- THPFunction's deleter calls cdata (a PyFunction)'s deleter. This deleter will trigger the deletion of the Edge (2) and trigger the chain of deletions.

What we would like here is to use the custom deleter with the PyFunction. However, the THPFunction object owns the PyFunction (and because the PyFunction is a member of THPFunction, when the THPFunction struct gets freed, the PyFunction can't exist by itself). So the solution I had was to make THPFunction own a `PyFunction*`.  The following steps happen now:
1) The use_count of (1) goes to 0
2) Py_DECREF is called on the THPFunction that owns the underlying PyFunction
3) THPFunction's refcount goes to 0
4) THPFunction's deleter is called
5) The PyFunction* is deleted with the custom deleter that either deletes the PyFunction on the spot OR queues it into the delete queue

Reply:
As I mentioned above I don't think it's necessary to add the custom deleter to PyFunctions because Python will start adding the THPFunctions that own the PyFunctions to a delete queue after only 50 of them.  I could undo the changes to PyFunction if we think that's good.

Reply:
Ok, thanks a lot for the detailed explanation! If we can then let's try to limit the number of indirections and keep the function as a member here.

Reply:
Will do!